### PR TITLE
Build against TF2.4

### LINKED ISSUE #2110: Switch to CUDA 11 and cuDNN 8
TF-Nightly has switched already and this should match TF2.4. We'll likely wait until 2.4rc0 to do this, but the groundwork was laid in #1950 and we can likely test both CUDA versions in CI.

https://groups.google.com/a/tensorflow.org/g/discuss/c/G4E2VPY-X_8
https://github.com/tensorflow/tensorflow/issues/42047

### PR DESCRIPTION
**Pull request by seanpmorgan**
# Description

Update TFA to build against TF2.4 

Fixes #2110

### CODE REVIEW

File: tensorflow_addons/utils/types.py [hunk 1]
```diff
@@ -50,6 +53,7 @@
     tf.Tensor,
     tf.SparseTensor,
     tf.Variable,
+    keras_tensor.KerasTensor,
```
Comment (seanpmorgan) [inline tensorflow_addons/utils/types.py:56]: This is to fix seq2seq testing. Unfortunetely it's not clear if a fix for https://github.com/tensorflow/tensorflow/issues/44613 will land in TF2.4 so it may be required to add this private import for a bit.

@qlzh727 any issue?

File: tensorflow_addons/text/tests/crf_test.py [hunk 2]
```diff
@@ -497,6 +497,9 @@ def test_tf_function():
     crf_decode(potentials, transition_params, sequence_length)
 
 
+@pytest.mark.skip(
+    "Resolve issue # -- tf.reverse_sequence doesn't support symbolic tensor on TF2.4"
+)
```
Comment (seanpmorgan) [inline tensorflow_addons/text/tests/crf_test.py:502]: https://github.com/tensorflow/tensorflow/issues/45136

File: tensorflow_addons/optimizers/tests/conditional_gradient_test.py [hunk 3]
```diff
@@ -32,10 +32,12 @@ def _dtypes_to_test(use_gpu):
     # The function "_DtypesToTest" is from
     #       "https://github.com/tensorflow/tensorflow/blob/5d4a6cee737a1dc6c20172a1dc1
     #        5df10def2df72/tensorflow/python/kernel_tests/conv_ops_3d_test.py#L53-L62"
+    #
+    #  Update cpu to use tf.half once issue in TF2.4 is fixed: https://github.com/tensorflow/tensorflow/issues/45136
     if use_gpu:
         return [tf.float32, tf.float64]
     else:
-        return [tf.half, tf.float32, tf.float64]
+        return [tf.float32, tf.float64]
```
Comment (seanpmorgan) [inline tensorflow_addons/optimizers/tests/conditional_gradient_test.py:40]: https://github.com/tensorflow/tensorflow/issues/45136

**Comment by seanpmorgan** [comment]
Heads up @yongtang we're having issues building custom-ops against TF2.4 for Windows. Started this SIG build thread:
https://groups.google.com/a/tensorflow.org/g/build/c/LbAw8RILvTg/m/ttnuhYU2BgAJ

**Comment by yongtang** [comment]
Thanks @seanpmorgan for the pointer. I think `--copt="/Zc:preprocessor"` will resolve the issue in Visual Studio on GitHub.Actions.

File: tensorflow_addons/layers/tests/crf_test.py [hunk 4]
```diff
@@ -284,7 +284,7 @@ def test_numerical_accuracy():
     base_model = tf.keras.Model(x_input, crf_outputs)
     model = ModelWithCRFLoss(base_model)
 
-    model.compile()
+    model.compile(optimizer="Adam")
```
Comment (WindQAQ) [inline tensorflow_addons/layers/tests/crf_test.py:287]: @seanpmorgan why we should specify adam here?

Comment (seanpmorgan) [reply tensorflow_addons/layers/tests/crf_test.py:287]: The default optimizer loss scaler had issue with mixed precision. Not related to our repo I believe.

File: tensorflow_addons/text/tests/crf_test.py [hunk 5]
```diff
@@ -497,6 +497,9 @@ def test_tf_function():
     crf_decode(potentials, transition_params, sequence_length)
 
 
+@pytest.mark.skip(
+    "Resolve issue # -- tf.reverse_sequence doesn't support symbolic tensor on TF2.4"
```
Comment (WindQAQ) [inline tensorflow_addons/text/tests/crf_test.py:501]: https://github.com/tensorflow/tensorflow/pull/45534 for tracking.

**Comment by satishpasumarthi** [comment]
Hi, when is this PR expected to be merged?

**Comment by seanpmorgan** [comment]
> Hi, when is this PR expected to be merged?

Stable TF 2.4 just released an hour ago, so will update PR and then we can merge

**Comment by bhack** [comment]
@seanpmorgan It doesn't build cause https://github.com/tensorflow/addons/pull/2238/ check https://github.com/tensorflow/addons/pull/2238/files#diff-347b0ae22433de298d1f5f70dd47844f85d0b05fc3d970ccdec0a8c6d8e4d463R66

**Comment by fsx950223** [comment]
> /cc @fsx950223 Do you have a already a fix for your todo at https://github.com/tensorflow/addons/pull/2238/files#diff-347b0ae22433de298d1f5f70dd47844f85d0b05fc3d970ccdec0a8c6d8e4d463R66 ?

I fixed it in #2291

File: tensorflow_addons/text/crf.py [hunk 6]
```diff
@@ -534,6 +535,10 @@ def crf_decode(
                   Contains the highest scoring tag indices.
       best_score: A [batch_size] vector, containing the score of `decode_tags`.
     """
+    if tf.__version__[:3] == "2.4":
+        warnings.warn(
+            "CRF Cecoding does may work in TF2.4. The bug has since been fixed in tensorflow/tensorflow##45534"
+        )
```
Comment (seanpmorgan) [inline tensorflow_addons/text/crf.py:541]: I don't see anyway around this unfortunately.

**Review by seanpmorgan — COMMENTED** [review]
@tensorflow/sig-addons-maintainers ready for review

File: tensorflow_addons/text/crf.py (hunk 6 above)
Comment (guillaumekln) [reply tensorflow_addons/text/crf.py:541]: If I understand correctly it is an issue for `KerasTensor`. Is it possible to check this tensor type and only raise an error for this case?

Comment (seanpmorgan) [reply tensorflow_addons/text/crf.py]: So KerasTensor is a private class and we'd have to  add crf.py to the exception list that we're not trying to grow:
https://github.com/tensorflow/tensorflow/issues/44613

I've updated the warning a bit, and since this will quickly throw an error if they are indeed KerasTensors I'm alright with it if everyone else is. Agree this isn't the best UX, but I also don't know why the fix didn't get cherry-picked into TF2.4

**Review by bhack — APPROVED** [review]

File: tensorflow_addons/text/crf.py (hunk 6 above)
Comment (WindQAQ) [reply tensorflow_addons/text/crf.py]: The patch is not cherrypicked in TF2.4. I think the warning is `CRF Cecoding does may work in TF2.5. ...`?
